protocol/contracts/nexus/Shards.sol

```diff
@@ -0,0 +1,232 @@
+//                       @@&                             @@(                      
+//                       @@&                             @@(                      
+//                       @@&                         @@@ @@(                      
+//                 @@@@@@@@@@@@@@@                   @@@ @@(                      
+//                 @@    @@&    @@                @@@@@@ @@(                      
+//                 @@    @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@(                      
+//                 @@           @@                                                
+//                 @@           @@                                                
+//                 @@@@@@@@@@@@@@@   
+
+pragma solidity ^0.8.4;
+
+import "@openzeppelin/contracts/token/ERC1155/ERC1155.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/security/Pausable.sol";
+import "@openzeppelin/contracts/token/ERC1155/extensions/ERC1155Burnable.sol";
+import "@openzeppelin/contracts/token/ERC1155/extensions/ERC1155Supply.sol";
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+
+
+interface IRelic { 
+    function getRelicId(address _owner) external view returns (uint256);
+    function hasRelic(address _owner) external view returns (bool);
+    function ownerOf(uint256 tokenId) external view returns (address);
+    function balanceOf(address _owner) external view returns(uint256);
+}
+
+
+contract Shards is
+    ERC1155,
+    Ownable,
+    Pausable,
+    ERC1155Burnable,
+    ERC1155Supply,
+    ReentrancyGuard
+{
+    constructor() ERC1155("") {}
+
+    // whitelisted partners to mint Shards for users
+    mapping (address => bool) public partnerMinters;
+    // Reserved itemId for each partner PARTNER_ADDRESS => ( SHARD_ID => AUTH? )
+    mapping (address => mapping (uint256=>bool)) public partnerAllowedIds;
+    // URIs
+    mapping (uint256 => string) public tokenURIs;
+    // Recipes
+    mapping(uint256 => Recipe) public recipes;
+
+    // Relic.sol
+    IRelic private RELIC;
+
+    struct Recipe {
+        // recipe ID
+        uint16 id;
+        // required Shards input Ids
+        uint256[] requiredIds;
+        // required Shards input amounts
+        uint256[] requiredAmounts;
+        // required Shards input Ids
+        uint256[] rewardIds;
+        // required Shards output amounts
+        uint256[] rewardAmounts;
+    }
+
+    modifier isRelic{
+        require(msg.sender==address(RELIC));
+        _;
+    }
+
+    event Transmutation(address Templar, uint256 recipeId);
+    event MintedShard(address to, uint256 shardId, uint256 amount);
+
+    //------- External -------//
+
+    // called from Relic when transfering items from Templar wallet into Relic
+    function equipShard(address _ownerAddress, uint256[] memory _shardIds, uint256[] memory _amounts) external isRelic {
+        
+        _beforeTokenTransfer(msg.sender, _ownerAddress, address(RELIC), _shardIds, _amounts, "");
```

Comment:
This is already called within `ERC1155._safeBatchTransferFrom()`, why do you need to call it explicitly here too?